File: MK-52_ESP32/src/Receiver_FILE_K.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receivers.hpp"
#include "../functions/functions.h"

#define __DEBUG

using namespace MK52_Interpreter;

unsigned long Receiver_FILE_K::init( void *components[]) {
    #ifdef __DEBUG
    Serial.println( "Init FILE_K");
    #endif
    _tr = (Receiver_Text *)components[COMPONENT_RECEIVER_TEXT];
    return Receiver::init(components);
}

void Receiver_FILE_K::activate( uint8_t scancode, int8_t parent){
    #ifdef __DEBUG
    Serial.println( "Activating receiver FILE_K");
    #endif
    Receiver::activate( scancode, parent);
    _lcd->updateStatusFMODE( " K ");
    _mode = 1;
    if(!scancode) return;
    tick(scancode);
}

int Receiver_FILE_K::tick( uint8_t scancode){
    int return_value = COMPONENT_RECEIVER_FILE_N;
    int r = _completeSubentry(scancode);
    if( r < NO_CHANGE) return return_value;
    if( r <= 0) return NO_CHANGE;
    scancode = (uint8_t)r;
    switch( scancode){
        // Column 0
        case 1:
            return_value = COMPONENT_RECEIVER_FILE_F;
            break;
        case 3:
            return_value = COMPONENT_RECEIVER_FILE_A;
            break;
            
        default: // all other buttons do nothing - keep K-mode
           return NO_CHANGE;
    }
    _mode = 0;
    return return_value;
}

int Receiver_FILE_K::_completeSubentry( uint8_t scancode){
    int8_t r = (int)scancode;
    // switch( _mode){
    //     default:
    //         break;
    // }
    _mode = 0;
    return r;
}

File: MK-52_Simulator/Receivers/Receiver_FILE_K.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    public class Receiver_FILE_K: Receiver_FILE
    {
        public Receiver_FILE_K(MK52_Host parent)
            : base( parent)
        {
            Moniker = "FILE_K";
        }

        public override void activate(string prevReceiver)
        {
            base.activate(prevReceiver);
            LCD_Manager lm = _parent.getLCD();
            lm.updateStatusFMODE(" K ");
            lm.forcePaint();
        }

        public override byte tick(byte scancode)
        {
            RPN_Functions _rpnf = _parent.getFunctions();
            switch (scancode)
            {
                case 0: // keyboard inactive
                    return 0;

                // Column 0
                case 1:
                    return _rpnf.requestNextReceiver("FILE_F");
                case 3:
                    return _rpnf.requestNextReceiver("FILE_A");

                // Column 1 does nothing (for now)

                // Column 2
                case 9:
                    _rpnf.execute(RPN_Functions.FUNC_LOADALL);
                    return _rpnf.requestNextReceiver("AUTO_N");
                case 10:
                    return _rpnf.requestNextReceiver("File_Overwrite_All");
                case 12:
                    return _rpnf.requestNextReceiver("File_All");

                // Column 3 does nothing (for now)
                // Column 4 does nothing (for now)
                // Column 5 does nothing (for now)
                // Column 6 does nothing (for now)

                // Column 7
                case 32:
                    return _rpnf.requestNextReceiver("Prog_Erase", "PROG_N");

                case 33:
                    // Shutdown signal
                    _rpnf.requestNextReceiver("FILE_N");
                    return 33;

                default: // all other buttons do nothing, keeping K-mode
                    return 0;
            }
            //return _rpnf.requestNextReceiver("FILE_N");
        }
    }
}
